### PR TITLE
Display relational fields in the list view

### PR DESCRIPTION
**Pull request by HichamELBSI**
Signed-off-by: HichamELBSI <elabbassih@gmail.com>

### What does it do?

Display relational fields in the content manager list view.

### Why is it needed?

Describe the issue you are solving.

### Related issue(s)/PR(s)

Let us know if this is related to any issue/pull request

Capture d’écran 2020-11-23 à 23 00 02

### CODE REVIEW

**Comment by derrickmehaffy** [comment]
@HichamELBSI thank you for the picture <3

File: packages/strapi-plugin-content-manager/admin/src/components/CustomTable/Row.js [hunk 1]
```diff
@@ -115,16 +123,22 @@ function Row({ canCreate, canDelete, canUpdate, isBulkable, row, headers, goTo }
       )}
       {headers.map(({ key, name, fieldSchema: { type }, cellFormatter }) => {
         const isMedia = type === 'media';
+        const isRelation = type === 'relation';
```
Comment (soupette) [inline packages/strapi-plugin-content-manager/admin/src/components/CustomTable/Row.js:126]: Can you create a component for the `Td` in order to simplify the code?

`const Td = (props) => { }` (we might need to find a better naming).

File: packages/strapi-plugin-content-manager/admin/src/components/RelationPreviewList/index.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+import { useIntl } from 'react-intl';
+
+import { getTrad } from '../../utils';
+import { useContentTypeLayout } from '../../hooks';
+import { Truncate, Truncated } from '../CustomTable/styledComponents';
+import CountWrapper from './CountWrapper';
+
+const RelationPreviewList = ({ name, value }) => {
+  const { formatMessage } = useIntl();
+  const { contentType } = useContentTypeLayout();
```
Comment (soupette) [inline packages/strapi-plugin-content-manager/admin/src/components/RelationPreviewList/index.js:14]: Here I suggest doing the following update:

In the `useFetchContentTypeLayout/utils/formatLayout` hook you can add the mainField to the list layout:

```

const formatListLayoutWithMetas = obj => {
  const formatted = obj.layouts.list.reduce((acc, current) => {
    const fieldSchema = get(obj, ['attributes', current], {});
    let metadatas = get(obj, ['metadatas', current, 'list'], {});

    const type = fieldSchema.type;

    if (type === 'relation') {
      metadatas = {
        ...metadatas,
        mainField: get(obj, ['metadatas', current, 'edit', 'mainField'], 'id'),
      };
    }

    acc.push({ key: `__${current}_key__`, name: current, fieldSchema, metadatas });

    return acc;
  }, []);

  return formatted;
};
```

And then in the `List/reducer` you will need to perform the same logic:

```
case ON_CHANGE_LIST_HEADERS: {
  const {
    target: { name, value },
  } = action;

  if (!value) {
    const { metadatas, attributes } = state.contentType;
    let metas = metadatas[name].list;

    if (attributes[name].type === 'relation') {
      const mainField = metadatas[name].edit.mainField;

      metas = { ...metas, mainField };
    }

    drafState.displayedHeaders.push({
      name,
      fieldSchema: attributes[name],
      metadatas: metas,
      key: `__${name}_key__`,
    });
  } else {
    drafState.displayedHeaders = state.displayedHeaders.filter(
      header => header.name !== name
    );
  }
  break;
}
```


With this update you can remove the dependency to the useContentTypeLayout hook

`const RelationPreviewList = ({ metadatas: { mainField }, value)  => {}`

File: packages/strapi-plugin-content-manager/admin/src/components/RelationPreviewList/index.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+  const mainField = get(contentType, ['metadatas', name, 'edit', 'mainField'], '');
+
+  return Array.isArray(value) ? (
```
Comment (soupette) [inline packages/strapi-plugin-content-manager/admin/src/components/RelationPreviewList/index.js:17]: I am not sure this is the correct logic. I believe that you should use the type that you can get directly from the fieldSchema instead.
```
const isSingle = ['oneWay', 'oneToOne', 'manyToOne'].includes(type)

if (isSingle) {
  return <Truncate><Truncated>{value ? value[mainField] : '-'}</Truncated></Truncate>
```

File: packages/strapi-plugin-content-manager/admin/src/components/RelationPreviewList/index.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    <Truncate>
+      <Flex>
+        <CountWrapper>
+          <Count count={value.length} />
```
Comment (soupette) [inline packages/strapi-plugin-content-manager/admin/src/components/RelationPreviewList/index.js:21]:
```
const size = value ? value.length : 0

<Count count={size} />`
```

File: packages/strapi-plugin-content-manager/admin/src/components/RelationPreviewList/index.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+                ? 'containers.ListPage.items.plural'
+                : 'containers.ListPage.items.singular'
+            ),
+          })}
+        </Truncated>
+      </Flex>
+    </Truncate>
+  ) : (
+    <Truncate>
+      <Truncated>{value[mainField] || '-'}</Truncated>
```
Comment (soupette) [inline packages/strapi-plugin-content-manager/admin/src/components/RelationPreviewList/index.js:37]: `value ? value[mainField] : '-'`

**Review by soupette — CHANGES_REQUESTED** [review]

File: packages/strapi-plugin-content-manager/admin/src/components/CustomTable/Row.js (hunk 1 above)
Comment (HichamELBSI) [reply packages/strapi-plugin-content-manager/admin/src/components/CustomTable/Row.js:126]: Sure

File: packages/strapi-plugin-content-manager/admin/src/components/RelationPreviewList/index.js (hunk 3 above)
Comment (HichamELBSI) [reply packages/strapi-plugin-content-manager/admin/src/components/RelationPreviewList/index.js:17]: Thanks

File: packages/strapi-plugin-content-manager/admin/src/components/CustomTable/Row.js (hunk 1 above)
Comment (HichamELBSI) [reply packages/strapi-plugin-content-manager/admin/src/components/CustomTable/Row.js:126]: WDYT about `<RowCell ... />` ? @soupette

File: packages/strapi-plugin-content-manager/admin/src/containers/ListView/reducer.js [hunk 6]
```diff
@@ -88,10 +88,18 @@ const listViewReducer = (state = initialState, action) =>
 
         if (!value) {
           const { metadatas, attributes } = state.contentType;
+          let metas = metadatas[name].list;
+
+          if (attributes[name].type === 'relation') {
+            const mainField = metadatas[name].edit.mainField;
+            const { relationType } = attributes[name];
```
Comment (soupette) [inline packages/strapi-plugin-content-manager/admin/src/containers/ListView/reducer.js:95]: This is not needed you have the relationType information in the fieldSchema object.

File: packages/strapi-plugin-content-manager/admin/src/components/RelationPreviewList/index.js [hunk 7]
```diff
@@ -0,0 +1,52 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { Flex, Padded, Count } from '@buffetjs/core';
+import { useIntl } from 'react-intl';
+
+import { getTrad } from '../../utils';
+import { Truncate, Truncated } from '../CustomTable/styledComponents';
+import CountWrapper from './CountWrapper';
+
+const RelationPreviewList = ({ metadatas: { mainField, relationType }, value }) => {
```
Comment (soupette) [inline packages/strapi-plugin-content-manager/admin/src/components/RelationPreviewList/index.js:10]: `const RelationPreviewList = ({ metadatas: { mainField }, relationType, value }) => {}`

File: packages/strapi-plugin-content-manager/admin/src/components/RelationPreviewList/index.js [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    </Truncate>
+  );
+};
+
+RelationPreviewList.defaultProps = {
+  value: null,
+};
+
+RelationPreviewList.propTypes = {
+  metadatas: PropTypes.object.isRequired,
```
Comment (soupette) [inline packages/strapi-plugin-content-manager/admin/src/components/RelationPreviewList/index.js:48]:
```
metadatas: PropTypes.shape({ 
  mainField: PropTypes.string.isRequired
}).isRequired,
```

File: packages/strapi-plugin-content-manager/admin/src/components/CustomTable/RowCell.js [hunk 9]
```diff
@@ -0,0 +1,34 @@
+import React, { memo } from 'react';
+import PropTypes from 'prop-types';
+import MediaPreviewList from '../MediaPreviewList';
+import RelationPreviewList from '../RelationPreviewList';
+import { Truncate, Truncated } from './styledComponents';
+
+const RowCell = ({ metadatas, type, value }) => {
```
Comment (soupette) [inline packages/strapi-plugin-content-manager/admin/src/components/CustomTable/RowCell.js:7]: `const RowCell = ({ metadatas, type,  relationType, value }) => {}`

File: packages/strapi-plugin-content-manager/admin/src/components/CustomTable/RowCell.js [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+  if (type === 'media') {
+    return <MediaPreviewList files={value} />;
+  }
+
+  if (type === 'relation') {
+    return <RelationPreviewList metadatas={metadatas} value={value} />;
```
Comment (soupette) [inline packages/strapi-plugin-content-manager/admin/src/components/CustomTable/RowCell.js:13]: `return <RelationPreviewList metadatas={metadatas}  relationType={relationType} value={value} />;`

File: packages/strapi-plugin-content-manager/admin/src/components/CustomTable/Row.js [hunk 11]
```diff
@@ -113,21 +120,15 @@ function Row({ canCreate, canDelete, canUpdate, isBulkable, row, headers, goTo }
...
-            {cellFormatter && cellFormatter(row)}
-            {!isMedia && !cellFormatter && (
-              <Truncate>
-                <Truncated>{memoizedDisplayedValue(name, type)}</Truncated>
-              </Truncate>
-            )}
-          </td>
-        );
-      })}
+      {headers.map(({ key, name, fieldSchema: { type }, cellFormatter, metadatas }) => (
```
Comment (soupette) [inline packages/strapi-plugin-content-manager/admin/src/components/CustomTable/Row.js:123]: {headers.map(({ key, name, fieldSchema: { relationType, type }, cellFormatter, metadatas }) => ()`

File: packages/strapi-plugin-content-manager/admin/src/components/CustomTable/Row.js [hunk 12]
```diff
@@ -113,21 +120,15 @@ function Row({ canCreate, canDelete, canUpdate, isBulkable, row, headers, goTo }
...
+        <td key={key}>
+          {cellFormatter ? (
+            cellFormatter(row)
+          ) : (
+            <RowCell type={type} metadatas={metadatas} value={memoizedDisplayedValue(name, type)} />
```
Comment (soupette) [inline packages/strapi-plugin-content-manager/admin/src/components/CustomTable/Row.js:128]:
```
<RowCell type={type} metadatas={metadatas}  relationType={relationType} value={memoizedDisplayedValue(name, type)} />
```

**Review by soupette — CHANGES_REQUESTED** [review]

**Review by soupette — APPROVED** [review]
LGTM

**Pull request merged**